src/robot/utils/robottime.py

```diff
@@ -420,6 +421,8 @@ def get_timestamp(self, daysep='', daytimesep=' ', timesep=':', millissep='.'):
 
     # Seam for mocking
     def _get_epoch(self):
+        if os.getenv('SOURCE_DATE_EPOCH'):
+            return float(os.getenv('SOURCE_DATE_EPOCH'))
         return time.time()
```

Comment:
This is pretty performance critical code, timestamps are generated for each keyword, loop, etc. both when they start and end, and thus always using `os.getenv` isn't great. It would be better to have the `if` outside the `_get_epoch` like this:

```python
if 'SOURCE_DATE_EPOCH' in os.environ:
    def _get_epoch(self):
        return float(os.getenv('SOURCE_DATE_EPOCH'))
else:
    def _get_epoch(self):
        return time.time()

Reply:
> Could you @Apteryks still clarify the use case for this change a bit?

My specific use case was just fixing the timestamps embedded in the generated documentation so that RobotFramework (and its libraries) could be built reproducibly in GNU Guix. Another, more contrived example, could be the desire that two RobotFramework test runs produce exactly the same output.

Reply:
I've made the suggested change, and it expectedly broke the new `test_write_to_logger_with_source_date_epoch_set`; which I tried fixing by mocking the `TIMESTAMP_CACHE` global object:
```diff
modified   utest/output/test_logger.py
@@ -4,6 +4,7 @@ import time
 import unittest
 import unittest.mock
 
+import robot.utils.robottime
 from robot.utils.asserts import assert_equal, assert_true, assert_false
 from robot.utils.robottime import format_time
 
@@ -59,6 +60,8 @@ class TestLogger(unittest.TestCase):
                              logger.msg.timestamp))
 
     @unittest.mock.patch.dict(os.environ, {'SOURCE_DATE_EPOCH': "0"})
+    @unittest.mock.patch('robot.utils.robottime.TIMESTAMP_CACHE',
+                         robot.utils.robottime.TimestampCache())
     def test_write_to_logger_with_source_date_epoch_set(self):
         logger = LoggerMock(('Hello, SOURCE_DATE_EPOCH!', 'INFO'))
         self.logger.register_logger(logger)
```
But it didn't work.  Any suggestion?

Reply:
Yeah, `if 'SOURCE_DATE_EPOCH' in os.environ:` in my proposal is executed when the `robottime` module is initially imported and thus patching it isn't easy. The code is so trivial that perhaps we wouldn't need a test at all. There could be just a comment explaining why `SOURCE_DATE_EPOCH` is used with an URL pointing to the original issue.